core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -611,24 +885,56 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
-  def localLogOrException(topicPartition: TopicPartition): Log = {
-    getPartitionOrException(topicPartition).localLogOrException
+  // Returns the Log for the partition if it is online and has one, otherwise raises an exception.
+  // Equivalent to localDeferredOrOnlineLogOrException() when using ZooKeeper.
+  def localOnlineLogOrException(topicPartition: TopicPartition): Log = {
+    onlinePartitionOrException(topicPartition).localLogOrException
+  }
+
+  // Returns the Log for the partition if it is deferred or online and has one, otherwise raises an exception.
+  // Equivalent to localOnlineLogOrException() when using ZooKeeper.
+  def localDeferredOrOnlineLogOrException(topicPartition: TopicPartition): Log = {
```

Comment:
I didn't see a response to my previous comment, so I will repeat it here in case it was missed. I think we need to be clear about the contract when the partition is in the "deferred" state. It's unclear to me why we need to expose either the `Log` or `Partition` object at all in this state.  Keep in mind that in this patch when the state is "deferred," the leader/epoch information inside `Partition` is guaranteed to be out of date. It is only inside `applyDeferredMetadataChanges` that we apply the latest metadata to `Partition`. It may be safe to rely on the old state, but what is the point?

If ultimately we think this is the right implementation--that is if ultimately we just want to continue with the old state until the initialization/reloading has completed--then I'm not sure I see the benefit of mixing the deferred state into `allPartitions`. It would be simpler to collect the deferred entries in a separate map which would be applied at the end when the loading is complete. Then we could dump all of these "deferredOrOnline" APIs and potentially the duplication inside `applyDeferredMetadataChanges`.

Reply:
Metadata gets deferred when the broker is fenced and new metadata arrives.  If new metadata is arriving then that probably means we are heartbeating successfully, and that likely means we'll be caught up soon.

I doubt we would end up in some intermediate state for very long.  I mean, it's possible, but it's unlikely -- we would have to be fenced, then back in communication, then out of communication.  The most likely event by far is that we have intermediate states for a short period of time.

So maybe the best thing to do is to assume the worst as soon as anything defers -- i.e. we don't expose the partition when it has deferred metadata.